### PR TITLE
Aaron - LA154.Transport

### PR DESCRIPTION
**Pull request by aarony2j**
#344
Please see comments in last commit.

### CODE REVIEW

File: R/zchunk_LA154.Transport.R [hunk 1]
```diff
@@ -32,71 +31,215 @@ module_gcam.usa_LA154.Transport_DISABLED <- function(command, ...) {
...
+
+      # Calculate the state-wise percentages for each of EIA's sector/fuel combinations that is relevant for disaggregating
+      # nation-level transportation energy to the states
+
+      # This starting table is transportation energy consumption by GCAM region (and other variables)
+      # We will first subset this data for only the USA and values that are > 0 in the historical periods
+      L154.in_EJ_R_trn_m_sz_tech_F_Yh %>%
+        filter(year %in% HISTORICAL_YEARS, GCAM_region_ID == gcam.USA_CODE) %>% # Filter for the USA and for historical years only
+        filter(value != 0) %>% # Here any rows with value of 0 will be lost, even if other years of the same group are nonzero
+        # To get these rows back, the dataframe will be spread and gathered to reintroduce those rows with NAs, which will then be replaced with 0
```
Comment (bpbond) [inline R/zchunk_LA154.Transport.R]: Hmm. If this is a small dataset, OK, but if not, it's probably faster to use `tidyr::complete` for this.

File: R/zchunk_LA154.Transport.R [hunk 2]
```diff
@@ -32,71 +31,215 @@ module_gcam.usa_LA154.Transport_DISABLED <- function(command, ...) {
...
+        Transportation_energy_consumption
+
+      # From the full state database, state shares will be calculated based on relevant EIA sector and fuel combinations
+      # These shares will later be multipled by the transportation energy consumption data above
+      # We will create a list first, concatenating EIA-fuel and -sector, so as to selectively remove those pairs from the dataset
+      trnUCD_EIA_mapping %>%
+        mutate(fuel_sector = paste(EIA_fuel, EIA_sector)) %>%
+        .[["fuel_sector"]] %>%
+        unique() ->
+        list_fuel_sector
```
Comment (bpbond) [inline R/zchunk_LA154.Transport.R]: `list_fuel_sector <- paste(trnUCD_EIA_mapping$EIA_fuel, trnUCD_EIA_mapping$EIA_sector) %>% unique` might be clearer and more compact? Up to you.

File: R/zchunk_LA154.Transport.R [hunk 3]
```diff
@@ -32,71 +31,215 @@ module_gcam.usa_LA154.Transport_DISABLED <- function(command, ...) {
...
+
+      # Here is the state-level data for which to calculate state shares
+      # We will first filter for only relevant EIA-fuel and -sector pairs
+      L101.EIA_use_all_Bbtu %>%
+        filter(year %in% HISTORICAL_YEARS) %>% # Ensure within historical period
+        mutate(fuel_sector = paste(EIA_fuel, EIA_sector)) %>% # Create concatenate list in base dataframe to match the syntax of our list above
+        filter(fuel_sector %in% list_fuel_sector) %>% # Filtering for just EIA-fuel/sector pairs
+        rename(value_state = value) %>%
+        select(state, EIA_fuel, EIA_sector, sector, fuel, year, value_state) ->
+        EIA_transportation_state
```
Comment (bpbond) [inline R/zchunk_LA154.Transport.R]: Could collapse these two lines into `select(state, EIA_fuel, EIA_sector, sector, fuel, year, value_state = value)` I think

File: R/zchunk_LA154.Transport.R [hunk 4]
```diff
@@ -32,71 +31,215 @@ module_gcam.usa_LA154.Transport_DISABLED <- function(command, ...) {
...
+
+      # To calculate the state share, we need to calculate the national amount
+      EIA_transportation_state %>%
+        group_by(EIA_fuel, EIA_sector, sector, fuel, year) %>% # Dropping state
+        summarise(value_national = sum(value_state)) %>%
```
Comment (bpbond) [inline R/zchunk_LA154.Transport.R]: Nice job of clearly showing what's happening!

File: R/zchunk_LA154.Transport.R [hunk 5]
```diff
@@ -32,71 +31,215 @@ module_gcam.usa_LA154.Transport_DISABLED <- function(command, ...) {
...
+        select(state, UCD_sector, mode, size.class, UCD_technology, UCD_fuel, fuel, year, value) ->
+        L154.in_EJ_state_trn_m_sz_tech_F
+
+      # As a final step, aggregate by fuel and name the sector
+      # This creates the second of three output tables
+      L154.in_EJ_state_trn_m_sz_tech_F %>%
+        group_by(state, fuel, year) %>%
+        summarise(value = sum(value)) %>%
+        ungroup() %>%
+        mutate("sector" = "transportation") %>%
```
Comment (bpbond) [inline R/zchunk_LA154.Transport.R]: I'm surprised this runs. Is this intended to be a column rename?

File: R/zchunk_LA154.Transport.R [hunk 6]
```diff
@@ -32,71 +31,215 @@ module_gcam.usa_LA154.Transport_DISABLED <- function(command, ...) {
...
+        Pop_national
+
+      L100.Pop_thous_state %>%
+        left_join_error_no_match(Pop_national, by = "year") %>%
+        mutate(value_share = value / value_national) %>% # Creating state share based on population
+        select(state, year, value_share) ->
+        Pop_state_share
+
+      if(OLD_DATA_SYSTEM_BEHAVIOR) {
+        # Old code used the function, apportion_to_states, but did not match modes.
```
Comment (bpbond) [inline R/zchunk_LA154.Transport.R]: What does this mean, it didn't match modes? Please clarify. Good job catching this mistake!

File: R/zchunk_LA154.Transport.R [hunk 7]
```diff
@@ -32,71 +31,215 @@ module_gcam.usa_LA154.Transport_DISABLED <- function(command, ...) {
...
+          state_share_data
+
+        # This is the underlying code of the function, apportion_to_states
+        data_final <- state_share_data[  ] *
+          nation_data[ rep( 1:nrow( nation_data ), length.out = nrow( state_share_data ) ),  ]
+
+        data_final <- as_tibble(data_final) # Need to convert the dataframe back to a tibble
+
+        data_final %>%
+          filter(row_number() <= 51) %>%
```
Comment (bpbond) [inline R/zchunk_LA154.Transport.R]: Ugh. Fragile. `length(old_df_order)`?

File: R/zchunk_LA154.Transport.R [hunk 8]
```diff
@@ -32,71 +31,215 @@ module_gcam.usa_LA154.Transport_DISABLED <- function(command, ...) {
...
+          mutate(mode = "Cycle") %>%
+          mutate(state = old_df_order) ->
+          State_cycle
+
+        data_final %>%
+          filter(row_number() > 51) %>%
```
Comment (bpbond) [inline R/zchunk_LA154.Transport.R]: As above.

File: R/zchunk_LA154.Transport.R [hunk 9]
```diff
@@ -32,71 +31,215 @@ module_gcam.usa_LA154.Transport_DISABLED <- function(command, ...) {
...
+        L154.out_mpkm_R_trn_nonmotor_Yh %>%
+          rename(value_mode = value) %>%
+          filter(GCAM_region_ID == gcam.USA_CODE) %>%
+          # Number of rows will change by adding states, so left_join_error_no_match cannot be used
+          left_join(Pop_state_share, by = "year") %>%
+          mutate(value = value_mode * value_share) %>% # Apportioning across the modes using the share data
+          filter(year %in% HISTORICAL_YEARS) %>% # Ensuring within historical period
+          select(state, mode, year, value) %>%
+          mutate(year = as.integer(year)) ->
+          L154.out_mpkm_state_trn_nonmotor_Yh
```
Comment (bpbond) [inline R/zchunk_LA154.Transport.R]: How big a difference will this be from OLD_DATA_SYSTEM_BEHAVIOR?

File: inst/extdata/gcam-usa/trnUCD_EIA_mapping.csv [hunk 10]
```diff
@@ -1 +1,22 @@
...
LDV_4W,refined liquids,AC,MG
Rail,coal,AC,DF
Rail,electricity,AC,ES
Rail,refined liquids,AC,DF
Ship Domestic,refined liquids,AC,RF
Ship International,refined liquids,AC,RF
Truck,refined liquids,AC,DF
Truck,gas,AC,DF
\ No newline at end of file
+# File: trnUCD_EIA_mapping.csv,,,
```
Comment (bpbond) [inline inst/extdata/gcam-usa/trnUCD_EIA_mapping.csv]: Remove ugly end-of-line Excel commas please

File: R/zchunk_LA154.Transport.R (hunk 1 above)
Comment (aarony2j) [inline R/zchunk_LA154.Transport.R]: @bpbond I tried but wasn't able to get it to work. If I included all the column names, then I got every combination possible. I tried using a vector with c() (as I saw online), but it caused an error. If you had any suggestions, I can try again.

Yes, this is a small dataset (~800x9).

File: R/zchunk_LA154.Transport.R (hunk 5 above)
Comment (aarony2j) [inline R/zchunk_LA154.Transport.R]: This is actually creating a column named "sector" and filling it with "transportation" entries. I took the quotes off "sector" and made a comment.

File: R/zchunk_LA154.Transport.R (hunk 6 above)
Comment (aarony2j) [inline R/zchunk_LA154.Transport.R]: matching the modes of transportation. I expanded the notes in the code to further explain.

File: R/zchunk_LA154.Transport.R (hunk 9 above)
Comment (aarony2j) [inline R/zchunk_LA154.Transport.R]: The sum is the same, as just the allocation was applied wrongly. For every other state, the values for the two modes of transportation are switched. The remainder states were correct.

File: R/zchunk_LA154.Transport.R (hunk 1 above)
Comment (bpbond) [inline R/zchunk_LA154.Transport.R]: You want to use `nesting()` inside the `complete` call, which means "only the combinations that appear in the data". That said, given that this is small and so reshaping is presumably fast, feel free to leave as-is too.

Comment (aarony2j) [inline R/zchunk_LA154.Transport.R]: Got it. Thanks!

File: R/zchunk_LA154.Transport.R (hunk 4 above)
Comment (gokuliyer) [inline R/zchunk_LA154.Transport.R]: Agreed!

**Comment by gokuliyer** [comment]
@ Ben: I think this is good to go.

**Pull request merged**